src/prep.py

```diff
@@ -0,0 +1,331 @@
+import argparse
+import os
+from pathlib import Path
+import re
+import shutil
+import subprocess
+import sys
+import textwrap
+from typing import Optional
+
+from icecream import ic
+
+try:
+    from .cli import (
+        error,
+        info,
+        success,
+        warn,
+        todo,
+        validate_path_from_cli,
+        validate_target_from_cli,
+        validate_build_cmd_from_cli,
+    )
+    from .types import Command
+except ImportError:
+    """
+    This is necessary to safeguard the user from running this script directly
+    which will generate errors due to relative importing rules in Python
+    """
+    print(
+        f"You should not run {__file__} as a standalone script.\n"
+        f"Use `$ python3 reduce.py {Path(__file__).stem}` instead!",
+        file=sys.stderr,
+    )
+    sys.exit(1)
+
+
+DEBUG = False
+ic.disable()
+
+
+def make_dot_i_file(
+    build_cmd: Command,
+    target: Path,
+    path_to_linux: Path,
+    output_dir: Path,
+    force_rm_existing_target: bool,
+) -> str:
+    """
+    Make the preprocessed .i file which can then be compiled sans build system
+    """
+
+    if (abs_target := path_to_linux / target).exists():
+        double_check_removal_with_user(
+            file=abs_target, force_rm=force_rm_existing_target
+        )
+
+    info(f"Making {target}...")
+    make_output = subprocess.run(
+        [*build_cmd, target], check=False, capture_output=True, cwd=path_to_linux
+    )
+    ic(make_output)
+
+    if make_output.returncode == 0:
+        info(f"Moving generated {target} file to {output_dir}")
+        shutil.move(path_to_linux / target, output_dir / target.name)
+        success(f"Successfully generated {target.name}")
+
+    return make_output.stdout.decode("utf-8")
+
+
+def get_compiler_invocation(target: Path, make_stdout: str) -> Command:
+    """
+    Parse out the CC invocation from make's raw output with V=1 on
+    """
+    pattern = f".*-o {target}.*"
+    matches = re.findall(pattern, make_stdout)
+    ic(matches)
+    assert (
+        len(matches) == 1
+    ), "Too many (or too few) compiler invocation matches!\n\
+            (not your fault, it's prepreduce's fault)"
+
+    main_invocation = matches[0].split()
+    main_invocation.insert(-1, "-c")
+    return clean_compiler_invocation(main_invocation)
+
+
+def clean_compiler_invocation(cc_invocation: Command) -> Command:
+    """
+    Remove all preprocessor flags as well as any relative pathing in invocation
+    """
+
+    to_remove = ("-I", "-D", "-Wp", "-include", "-Werror", "./", "-U", "-E")
+    rules = [(lambda rule: lambda flag: rule not in flag)(rule) for rule in to_remove]
```

Comment:
should `rule not in flag` instead be `not flag.startswith(rule)` rather than scan the entire flag?

Reply:
`in` should stop once it finds it (presumably at the beginning) and not proceed to scan the whole string.

I think `in` is better than `startswith` because I believe I've seen a flag like `"./lib/..."` (with quotes present) which means `./` wouldn't be the start of the flag.

Can you confirm if you've seen a flag wrapped in quotes like that or perhaps it was some part of my processing of the flags that added a quote and that's what I'm remembering.

Reply:
The problem I have with `in` is when it doesn't find the needle in the haystack.  startswith will not look further into the haystack then the length of the needle.

Reply:
> I think in is better than startswith because I believe I've seen a flag like "./lib/..." (with quotes present) which means ./ wouldn't be the start of the flag.

Ah, sorry I missed your point about this.

Do command line args beginning with `./` exist BOTH quoted and unquoted?

Taking a step back, `./some/path` doesn't make sense to remove in isolation.

Some flags are pairs, like `-o` and whatever follows it.  In such cases we should be removing the flag AND whatever follows it.

As such, I suspect that flags like `./some/path` (regardless of whether they are quoted or not) aren't really position independent flags like `-c` is.  As such, removing them one at a time might miss removing the flag that comes before them.

In this case I suspect that `./some/path` flags follow `-I` or `-include`.  As such, we should treat those (and `-D` and `-U`) with more care.  We need to handle those differently.  In particular, `-I`, `-D`, and `-U` may or may not be space delimated from their arguments.

`-Dfoo` is equivalent to `-D foo`. So if we split on spaces, we need to check if the flag that starts with `-D` has a space immediately after the `D`; if so remove `-D` and the following flag, if not then remove the entire flag.  Same for the above.

Of the current flags in `to_remove`, it only makes sense to remove `-E` and `-Werror` one by one, as they don't imply any positional relationship to other flags.

Reply:
Got it, this requires a semi-substantial refactor of the current flag-cleaning processes.

Will think it out.